Autohide feature.

pelatx opened the pull request:
Hi.

This PR adds the AutoHide feature, similar to the one described in the old issue #17.

This works by adding an entry in the preferences context menu to enable/disable the feature.

When enabled, clicking on the editor results in hidding the Outline List.
To unhide it relies on the default behavior of the extension: click on the button on main toolbar.

Greetings.

Hirse commented on src/Preferences.js:
```diff
@@ -65,6 +65,15 @@ define(function (require, exports, module) {
+        }, {
+            id: "autohide",
+            type: "boolean",
+            value: true,
```
The default value should be false.

pelatx replied:
I was doubting about that.

Hirse commented on src/Preferences.js:
```diff
@@ -65,6 +65,15 @@ define(function (require, exports, module) {
+                name: Strings.PREF_AUTOHIDE_NAME,
+                description: Strings.PREF_AUTOHIDE_DESC
+            },
+            inContextMenu: true
```
I am not sure about putting this in the context menu.
It doesn't really belong to the other options. Maybe with a separator above.

pelatx replied:
Or with its own button?

If you think autohide is an interesting feature and has merge possibility, I have no problem trying to fix it.

Hirse replied:
There is an issue about it, so there seems to be demand, therefore I would absolutely merge it when it is ready.

Hirse replied:
The setting is a toggle, so having its own button might cause some design challenges.

pelatx replied:
Yes, a button for a toggle is a little bit strange on the panel.
Well, it should work like the button on the main toolbar, highlighted when autohide is enabled.
And it seems that it would not be suitable.

The idea of placing it in the View menu does not like me too much.
Because until now, your extension is self-contained in its own panel and its button in the main toolbar. And personally, I like it that way.

But as you say, maybe it's a reasonable starting point.

Hirse commented:
For hiding the outline, does it work to use the [`showOutline` function](https://github.com/Hirse/brackets-outline-list/blob/master/src/OutlineManager.js#L158) instead of toggling the `enabled` setting?
When the `enabled` is changed, the Outline will be re-generated, which is not strictly necessary.

Hirse commented:
That makes sense. 

If you look at the extension in the issue, it adds a placeholder area on the side that you can click on to show it. That would probably be the ideal solution.

pelatx commented:
Oh, I did not watch the video of that extension.

I agree that it seems the right thing to do. Working ...

pelatx commented:
@Hirse, I have been testing different things and the one I have found more comfortable to use and, moreover, easier to implement is a "pure" auto-hide. No clicks, just "hovering" the main toolbar to unhide and the editor area to hide.

You can see it in [this video](https://vimeo.com/208719851).

What do you think about it?

Hirse commented:
Looks cool. I think that could work.

One thing though, in the video it looks like the bottom status bar is disappearing when you hover-open the outline.

pelatx commented:
Yes, It disappears. But nothing to do with Outline List. My Brackets installation does this sort of things frequently. 

I have checked out a git branch containing version 1.0.1, without any of my last commits, and the same thing happens. But not always.

Maybe this is related to the Linux version. I have a Windows 10 virtual machine with Brackets installed and It works correctly.

I have updated the video of the link above. I have not written any line of code more since I recorded the first one, and It does not happen.

I've moved the toggle to the View menu. Also it seems that the Auto-hide works correctly. 

But I will not commit yet because I'm going to try a placeholder.
Without it, when No Distractions mode is activated, the hover area gets lost (not when in sidebar).

Hirse commented:
Thanks. Good thinking. 👍

pelatx commented:
Sorry for the last and too many commits to pass tests.

Auto-hide now uses `showOutline` and `hideOutline` functions to work.
Added placeholder when hiding (only when it is beside the toolbar).
It uses the `mouseenter` event now, instead of click.

Hirse commented on main.js:
```diff
@@ -37,6 +39,11 @@ define(function (require, exports, module) {
 
+    /* beautify preserve:start *//* eslint-disable no-multi-spaces */
+    var COMMAND_AUTOHIDE   = "outline.autohide";
+    var MENU_ITEM_AUTOHIDE = "Outline List Autohide";
```
Is this a string to be seen by the user? If so, it should be translatable.

pelatx replied:
Oh! I had it pending and I forgot.

Hirse commented on src/OutlineManager.js:
```diff
@@ -214,13 +214,82 @@ define(function (require, exports, module) {
+        $(".content").css("right", $placeHolder.width() + toolbarPx + "px");
+
+        $placeHolder.on("mouseenter", function () {
+            if (prefs.get("enabled")) {
```
Is this check really necessary?
If the outline was disabled, the placeholder shouldn't even be there, right?

pelatx replied:
I'll check it out

pelatx replied:
You are right. After trying several approaches and changing the code a lot, this has remained there by mistake.

Hirse commented on src/OutlineManager.js:
```diff
@@ -214,13 +214,82 @@ define(function (require, exports, module) {
+     * Enable/disable the Outline Auto-hide.
+     * @param {boolean} enable True to enable, false to disable.
+     */
+    function enableAutohide(enable) {
```
Please call it `toggleAutohide` as it can do both.

Hirse commented on src/Preferences.js:
```diff
@@ -65,6 +65,15 @@ define(function (require, exports, module) {
+                name: Strings.PREF_AUTOHIDE_NAME,
+                description: Strings.PREF_AUTOHIDE_DESC
+            },
+            inContextMenu: false
```
I think you can remove this. It isn't present in the other preferences where it is false.

Alternatively add it everywhere. I just like it consistent.

pelatx replied:
Ok.

Hirse commented on styles/styles.css:
```diff
@@ -102,6 +102,15 @@
 
+#outline-placeholder {
+    border-top: 1px solid rgba(255, 255, 255, 0.05);
+    text-shadow: 0 1px 2px rgba(0, 0, 0, 0.5);
```
Does the placeholder have text?

pelatx replied:
Ups! Copy&Paste error.

Hirse commented:
No problem, thanks for the changes.

I'll have to play around with it myself later.

pelatx commented:
Bad implementation after all. 

Do not waste time with this PR. If you're doing it, sorry about that.

If I can, I will do a full re-writing.

Hirse commented:
Don't worry about it. I am grateful that you spend the effort and time on making a contribution.

Btw, an update of the Spanish translation is highly appreciated, regardless of the state of this feature. 😉
